test/suite/test_prepare_hs03.py

```diff
@@ -0,0 +1,209 @@
+            # Each session will update many consecutive keys.
+            start = (j * nkeys)
+            end = start + nkeys
+            for i in range(start, end):
+                cursors[j].set_key(ds.key(nrows + i))
+                cursors[j].set_value(prepare_value)
+                self.assertEquals(cursors[j].insert(), 0)
+            sessions[j].prepare_transaction('prepare_timestamp=' + timestamp_str(4))
+
+        hs_writes = self.get_stat(stat.conn.cache_write_hs) - hs_writes_start
+
+        # Assert if not writing anything to the history store.
+        self.assertGreaterEqual(hs_writes, 0)
+
+        # Test if we can read prepared updates from the history store.
+        cursor = self.session.open_cursor(self.uri)
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(1))
+        for i in range(1, nsessions * nkeys):
+            cursor.set_key(ds.key(nrows + i))
+            # The search should pass.
+            self.assertEqual(cursor.search(), 0)
+            # Correctness Test - commit_value should be visible
+            self.assertEquals(cursor.get_value(), commit_value)
+            # Correctness Test - prepare_value should NOT be visible
+            self.assertNotEquals(cursor.get_value(), prepare_value)
+        cursor.close()
+
+        # Close all cursors and sessions, this will cause prepared updates to be
+        # rollback-ed
+        for j in range (0, nsessions):
+            cursors[j].close()
+            sessions[j].close()
+
+        # Corrupt the table, Call salvage to recover data from the corrupted table and call verify
+        self.corrupt_salvage_verify()
+
+        self.session.commit_transaction()
+        self.session.checkpoint()
+
+        # Corrupt the table, Call salvage to recover data from the corrupted table and call verify
+        self.corrupt_salvage_verify()
+
+        # Finally, search for the keys inserted with commit timestamp
+        cursor = self.session.open_cursor(self.uri)
+        self.pr('Read Keys')
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(1))
+        for i in range(1, nkeys):
+            cursor.set_key(ds.key(nrows + i))
+            # The search should pass
+            self.assertEqual(cursor.search(), 0)
+            # Correctness Test - commit_value should be visible
+            self.assertEquals(cursor.get_value(), commit_value)
+        cursor.close()
+
+        self.session.commit_transaction()
+        self.session.checkpoint()
+
+        # Simulate a crash by copying to a new directory(RESTART).
+        copy_wiredtiger_home(".", "RESTART")
+
+        # Open the new directory.
+        self.conn = self.setUpConnectionOpen("RESTART")
+        self.session = self.setUpSessionOpen(self.conn)
+        cursor = self.session.open_cursor(self.uri)
+
+        # Search the keys inserted with commit timestamp after crash
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(1))
```

Comment:
All the read timestamps must point to the latest timestamp. It is to make sure that even with the latest timestamp, we could only see the commit value only.

Reply:
Fixed